Lib/importlib/_bootstrap.py

```diff
@@ -1121,25 +1121,13 @@ def _setup(sys_module, _imp_module):
 
     # Directly load built-in modules needed during bootstrap.
     self_module = sys.modules[__name__]
-    for builtin_name in ('_warnings',):
+    for builtin_name in ('_thread', '_warnings', '_weakref'):
         if builtin_name not in sys.modules:
             builtin_module = _builtin_from_name(builtin_name)
         else:
             builtin_module = sys.modules[builtin_name]
         setattr(self_module, builtin_name, builtin_module)
 
-    # Directly load the _thread module (needed during bootstrap).
```

Comment:
Please look also at `_bootstrap_external.py`. It contains almost duplicated code.

Reply:
Looks like you're right. @ericsnowcurrently, out of curiosity, why the two "bootstrap" modules?

Reply:
A while back we split the fundamental import machinery from the file-based parts.  The "external" part encapsulates the file-based finders and loaders (e.g. .py, extensions) and related functionality.  In addition to maintenance benefits, during runtime startup it helps to have a hard separation between the two.  Furthermore, the split also supports some planned work (encapsulating the import state).

Reply:
In particular, with the multi-phase bootstrap refactoring, the builtin and frozen module support is initialised as one of the last steps in the core runtime initialisation (https://github.com/python/cpython/blob/master/Python/pylifecycle.c#L717), while setting up filesystem imports doesn't happen until the main interpreter is being configured (https://github.com/python/cpython/blob/master/Python/pylifecycle.c#L803)

Reply:
Thank you both for the explanation. I think if those files had been named something else (such as `_bootstrap_core.py` and `_bootstrap_importers.py`) the situation would have been less confusing to me ("externals" sounded it was something else entirely). Also, you could consider improving the top-level module docstrings and/or comments to describe the split accurately.